File: cmd/quadlsp/tests/test_lsp.cc
#include <cassert>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>

// Simple test to verify LSP executable works correctly
// This is more of an integration test

static bool runLspTest(const std::string& input, const std::string& expectedSubstring) {
	// Create a temporary file with the input
	FILE* tmpInput = tmpfile();
	if (!tmpInput) {
		std::cerr << "Failed to create temp file" << std::endl;
		return false;
	}

	fwrite(input.c_str(), 1, input.length(), tmpInput);
	rewind(tmpInput);

	// Run the LSP server with input from temp file
	std::string command = "build/debug/cmd/quadlsp/quadlsp < /dev/stdin";
	FILE* pipe = popen(command.c_str(), "r");
	if (!pipe) {
		fclose(tmpInput);
		return false;
	}

	// Read output
	char buffer[4096];
	std::string output;
	while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
		output += buffer;
	}

	int status = pclose(pipe);
	fclose(tmpInput);

	// Check if expected substring is in output
	bool found = output.find(expectedSubstring) != std::string::npos;
	if (!found) {
		std::cerr << "Expected substring not found: " << expectedSubstring << std::endl;
		std::cerr << "Got output: " << output.substr(0, 200) << "..." << std::endl;
	}

	return found;
}

static void testInitialize() {
	std::cout << "Testing initialize..." << std::endl;

	std::string request = R"(Content-Length: 113

{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"capabilities":{},"rootUri":"file:///tmp"}})";

	// Should receive a response with capabilities
	assert(runLspTest(request, "capabilities"));
	assert(runLspTest(request, "textDocumentSync"));
	assert(runLspTest(request, "quadlsp"));

	std::cout << "✓ Initialize test passed" << std::endl;
}

static void testShutdown() {
	std::cout << "Testing shutdown..." << std::endl;

	std::string request = R"(Content-Length: 58

{"jsonrpc":"2.0","id":2,"method":"shutdown","params":{}})";

	// Should receive null result
	assert(runLspTest(request, "\"result\":null"));

	std::cout << "✓ Shutdown test passed" << std::endl;
}

static void testCompletion() {
	std::cout << "Testing completion..." << std::endl;

	std::string request =
			R"(Content-Length: 150

{"jsonrpc":"2.0","id":3,"method":"textDocument/completion","params":{"textDocument":{"uri":"file:///tmp/test.qd"},"position":{"line":0,"character":0}}})";

	// Should receive completion items
	assert(runLspTest(request, "\"label\":\"add\""));
	assert(runLspTest(request, "\"label\":\"sub\""));
	assert(runLspTest(request, "\"label\":\"mul\""));
	assert(runLspTest(request, "\"kind\":3"));

	std::cout << "✓ Completion test passed" << std::endl;
}

static void testJsonValidation() {
	std::cout << "Testing JSON format..." << std::endl;

	// Test that initialize returns valid JSON structure
	std::string request = R"(Content-Length: 113

{"jsonrpc":"2.0","id":1,"method":"initialize","params":{"capabilities":{},"rootUri":"file:///tmp"}})";

	// Check for proper JSON structure markers
	assert(runLspTest(request, "\"jsonrpc\":\"2.0\""));
	assert(runLspTest(request, "\"id\":1"));
	assert(runLspTest(request, "\"result\":{"));

	std::cout << "✓ JSON validation test passed" << std::endl;
}

static void testDocumentSymbols() {
	std::cout << "Testing document symbols..." << std::endl;

	std::string request =
			R"(Content-Length: 139

{"jsonrpc":"2.0","id":4,"method":"textDocument/documentSymbol","params":{"textDocument":{"uri":"file:///tmp/test_symbols.qd"}}})";

	// Should receive symbol list
	assert(runLspTest(request, "\"name\":"));
	assert(runLspTest(request, "\"kind\":12"));

	std::cout << "✓ Document symbols test passed" << std::endl;
}

int main() {
	std::cout << "=== Running LSP Tests ===" << std::endl;

	try {
		testJsonValidation();
		testInitialize();
		testShutdown();
		testCompletion();
		testDocumentSymbols();

		std::cout << "\n✅ All LSP tests passed!" << std::endl;
		return 0;
	} catch (const std::exception& e) {
		std::cerr << "❌ Test failed with exception: " << e.what() << std::endl;
		return 1;
	} catch (...) {
		std::cerr << "❌ Test failed with unknown exception" << std::endl;
		return 1;
	}
}

File: cmd/quadlsp/tests/test_lsp.py
#!/usr/bin/env python3
"""
Test suite for Quadrate Language Server Protocol (LSP)
"""

import json
import subprocess
import sys
from pathlib import Path

class LSPTester:
    def __init__(self, lsp_path):
        self.lsp_path = lsp_path
        self.test_count = 0
        self.passed = 0
        self.failed = 0

    def send_request(self, request_dict):
        """Send a JSON-RPC request to the LSP server and get response"""
        request_json = json.dumps(request_dict)
        message = f"Content-Length: {len(request_json)}\r\n\r\n{request_json}"

        try:
            result = subprocess.run(
                [self.lsp_path],
                input=message.encode(),
                capture_output=True,
                timeout=2
            )

            # Parse response
            output = result.stdout.decode()
            if not output:
                return None

            # Skip Content-Length header
            lines = output.split('\r\n')
            for i, line in enumerate(lines):
                if line == '':
                    json_str = '\r\n'.join(lines[i+1:])
                    return json.loads(json_str)

            return None
        except subprocess.TimeoutExpired:
            print("⏱️  Request timed out")
            return None
        except json.JSONDecodeError as e:
            print(f"❌ JSON decode error: {e}")
            print(f"   Output: {output[:200]}")
            return None

    def assert_equal(self, actual, expected, test_name):
        """Assert that actual equals expected"""
        self.test_count += 1
        if actual == expected:
            self.passed += 1
            print(f"✓ {test_name}")
            return True
        else:
            self.failed += 1
            print(f"✗ {test_name}")
            print(f"  Expected: {expected}")
            print(f"  Got: {actual}")
            return False

    def assert_contains(self, container, item, test_name):
        """Assert that item is in container"""
        self.test_count += 1
        if item in container:
            self.passed += 1
            print(f"✓ {test_name}")
            return True
        else:
            self.failed += 1
            print(f"✗ {test_name}")
            print(f"  Expected '{item}' to be in: {container}")
            return False

    def test_initialize(self):
        """Test LSP initialize request"""
        print("\n=== Testing Initialize ===")

        request = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {
                "capabilities": {},
                "rootUri": "file:///tmp"
            }
        }

        response = self.send_request(request)
        if not response:
            print("❌ No response received")
            self.failed += 1
            return

        self.assert_equal(response.get("jsonrpc"), "2.0", "Initialize: JSON-RPC version")
        self.assert_equal(response.get("id"), 1, "Initialize: Request ID")

        result = response.get("result", {})
        capabilities = result.get("capabilities", {})

        self.assert_equal(capabilities.get("textDocumentSync"), 1, "Initialize: Text sync capability")
        self.assert_equal(capabilities.get("documentFormattingProvider"), True, "Initialize: Formatting capability")
        self.assert_contains(capabilities, "completionProvider", "Initialize: Completion capability")

        serverInfo = result.get("serverInfo", {})
        self.assert_equal(serverInfo.get("name"), "quadlsp", "Initialize: Server name")
        self.assert_equal(serverInfo.get("version"), "0.1.0", "Initialize: Server version")

    def test_shutdown(self):
        """Test LSP shutdown request"""
        print("\n=== Testing Shutdown ===")

        request = {
            "jsonrpc": "2.0",
            "id": 2,
            "method": "shutdown",
            "params": {}
        }

        response = self.send_request(request)
        if not response:
            print("❌ No response received")
            self.failed += 1
            return

        self.assert_equal(response.get("jsonrpc"), "2.0", "Shutdown: JSON-RPC version")
        self.assert_equal(response.get("id"), 2, "Shutdown: Request ID")
        self.assert_equal(response.get("result"), None, "Shutdown: Null result")

    def test_completion(self):
        """Test LSP completion request"""
        print("\n=== Testing Completion ===")

        request = {
            "jsonrpc": "2.0",
            "id": 3,
            "method": "textDocument/completion",
            "params": {
                "textDocument": {
                    "uri": "file:///tmp/test.qd"
                },
                "position": {
                    "line": 0,
                    "character": 0
                }
            }
        }

        response = self.send_request(request)
        if not response:
            print("❌ No response received")
            self.failed += 1
            return

        self.assert_equal(response.get("jsonrpc"), "2.0", "Completion: JSON-RPC version")
        self.assert_equal(response.get("id"), 3, "Completion: Request ID")

        result = response.get("result", {})
        items = result.get("items", [])

        # Check that we have completion items
        if len(items) > 0:
            self.passed += 1
            print(f"✓ Completion: Has items ({len(items)} items)")
        else:
            self.failed += 1
            print("✗ Completion: No items returned")

        # Check for specific instructions
        labels = [item.get("label") for item in items]
        for instruction in ["add", "sub", "mul", "dup", "swap"]:
            self.assert_contains(labels, instruction, f"Completion: Has '{instruction}'")

        # Check item structure
        if items:
            first_item = items[0]
            self.assert_contains(first_item, "label", "Completion: Item has label")
            self.assert_equal(first_item.get("kind"), 3, "Completion: Item kind is Function")

    def test_diagnostics(self):
        """Test LSP diagnostics (error detection)"""
        print("\n=== Testing Diagnostics ===")

        # Send didOpen with invalid code
        request = {
            "jsonrpc": "2.0",
            "method": "textDocument/didOpen",
            "params": {
                "textDocument": {
                    "uri": "file:///tmp/invalid.qd",
                    "languageId": "quadrate",
                    "version": 1,
                    "text": "fn invalid( -- ) {\n  this is not valid quadrate code\n}\n"
                }
            }
        }

        response = self.send_request(request)
        # Note: didOpen is a notification, so no response expected
        # In a real test, we'd need to read the publishDiagnostics notification
        # For now, just verify the request doesn't crash the server
        print("✓ Diagnostics: Server handles didOpen without crashing")
        self.passed += 1
        self.test_count += 1

    def run_all_tests(self):
        """Run all tests"""
        print("=" * 50)
        print("Quadrate LSP Test Suite")
        print("=" * 50)

        self.test_initialize()
        self.test_shutdown()
        self.test_completion()
        self.test_diagnostics()

        print("\n" + "=" * 50)
        print(f"Tests run: {self.test_count}")
        print(f"Passed: {self.passed}")
        print(f"Failed: {self.failed}")
        print("=" * 50)

        if self.failed == 0:
            print("\n✅ All tests passed!")
            return 0
        else:
            print(f"\n❌ {self.failed} test(s) failed")
            return 1

def main():
    # Find LSP executable - check multiple possible locations
    possible_paths = [
        Path("build/debug/cmd/quadlsp/quadlsp"),  # From project root
        Path("build/release/cmd/quadlsp/quadlsp"),  # From project root
        Path("cmd/quadlsp/quadlsp"),  # From build directory (when run by meson)
        Path("../../../cmd/quadlsp/quadlsp"),  # Relative from test location
    ]

    lsp_path = None
    for path in possible_paths:
        if path.exists():
            lsp_path = path
            break

    if not lsp_path:
        print("❌ LSP executable not found. Please build the project first.")
        print("   Searched paths:")
        for p in possible_paths:
            print(f"     - {p.absolute()}")
        return 1

    print(f"Using LSP: {lsp_path}\n")

    tester = LSPTester(str(lsp_path))
    return tester.run_all_tests()

if __name__ == "__main__":
    sys.exit(main())
